_posts/2022-09-28-ltr-with-opensearch-and-metarank.markdown

```diff
@@ -0,0 +1,268 @@
+        {"name": "availability", "value": true}
+  ]
+}
+```
+2. **Impression**: visitor viewed an item listing. Example: search results, collection, recommendation widget was displayed.
+```json
+{
+  "event": "ranking",
+  "id": "81f46c34-a4bb-469c-8708-f8127cd67d27",
+  "timestamp": "1599391467000",
+  "user": "user1",
+  "session": "session1",
+  "fields": [
+      {"name": "query", "value": "socks"}
+  ],
+  "items": [
+        {"id": "item3", "relevancy":  2.0},
+        {"id": "item1", "relevancy":  1.0},
+        {"id": "item2", "relevancy":  0.5} 
+  ]
+}
+```
+3. **Interaction**: visitor acted on an item from the list. Example: click, add-to-cart, mouse hover, like.
+```json
+{
+  "event": "interaction",
+  "id": "0f4c0036-04fb-4409-b2c6-7163a59f6b7d",
+  "impression": "81f46c34-a4bb-469c-8708-f8127cd67d27",
+  "timestamp": "1599391467000",
+  "user": "user1",
+  "session": "session1",
+  "type": "purchase",
+  "item": "item1",
+  "fields": [
+        {"name": "count", "value": 1},
+        {"name": "shipping", "value": "DHL"}
+  ],
+}	
+```
+
+### Mapping events to ranking factors
+
+Metarank has a rich DSL, which can help you map feedback events to the actual numerical ranking factors. Apart from some simple cases like [directly using item metadata numerical fields](https://docs.metarank.ai/reference/overview/feature-extractors/scalar), Metarank can also handle complex ML features.
+
+You can do a one-hot/label encoding of low-cardinality string fields:
+```yaml
+- name: genre
+  type: string
+  scope: item
+  source: item.genre
+  values:
+  - comedy
+  - drama
+  - action
+```
+
+Extract and one-hot encode a mobile/desktop/tablet category from a User-Agent field:
+```yaml
+- name: platform
+  type: ua
+  field: platform
+  source: ranking.ua
+```
+
+A sliding window count of interaction events for a particular item:
+```yaml
+- name: item_click_count
+  type: window_count
+  interaction: click
+  scope: item
+  bucket_size: 24h         // make a counter for each 24h rolling window
+  windows: [7, 14, 30, 60] // on each refresh, aggregate to 1-2-4-8 week counts
+  refresh: 1h
+```
+
+Rates:
+```yaml
+- name: CTR
+  type: rate
+  top: click      // divide number of clicks
+  bottom: impression // to number of examine events
+  scope: item
+  bucket: 24h     // aggregate over 24-hour buckets
+  periods: [7, 14, 30, 60] // sum buckets for multiple time ranges
+```
+
+There are many more feature extractors available, so check [Metarank docs](https://docs.metarank.ai) for other examples and a more detailed [quick-start guide](https://docs.metarank.ai/introduction/quickstart).
+
+These feature extractors map events into numerical features and form a set of implicit judgments, which are later used for ML model training:
+
+![Implicit judgments diagram]({{ site.baseurl }}/assets/media/blog-images/2022-09-28-ltr-with-opensearch-and-metarank/ij.png){:class="img-centered"}
+
+## Sending requests
```

Comment:
Do you have any metrics on the search latency from user side? I mean from sending the OpenSearch request to getting the metarank rerank results. I am not sure what's the ratio of latency increasing comparing to the original OpenSearch query.

Reply:
Added an extra paragraph and a diagram about re-scoring latency.